src/briefcase/commands/run.py

```diff
@@ -1,23 +1,273 @@
+            returning a "clean" line without any log system preamble.
+        :param clean_output: Should the output displayed to the user be the
+            "clean" output? (Default: True).
+        :param success_filter: A function that will operate on a string
+            containing the last 10 lines of "clean" (i.e., preamble filtered)
+            logs, returning True if a "success" condition has been detected. If
+            the success filter returns True, the log process will be terminated.
+        :param failure_filter: A function that will operate on a string
+            containing the last 10 lines of "clean" (i.e., preamble filtered)
+            logs, returning True if a "failure" condition has been detected. If
+            the failure filter returns True, the log process will be terminated.
+        """
+        self.log_popen = log_popen
+        self.success = None
+        self.clean_filter = clean_filter
+        self.clean_output = clean_output
+
+        self.recent_history = []
+        self.success_filter = success_filter
+        self.failure_filter = failure_filter
+
+    def __call__(self, line):
+        """Filter a single line of a log.
+
+        :param line: A single line of raw system log content, including the newline.
+        """
+        # Compute the clean line
+        if self.clean_filter:
+            filtered = self.clean_filter(line)
+
+            # If the clean filter says the line can be dumped, return None
+            if filtered is None:
+                return None
+
+            # If there's a cleaned line, we can determine if it should be included in analysis
+            clean_line, included = filtered
+        else:
+            # If we don't perform cleaning, we assume all content is potentially
+            # Python, and should be included in analysis
+            clean_line = line
+            display_line = line
+            included = True
+
+        # If we're not using clean output, use the raw line for display.
+        if self.clean_output:
+            display_line = clean_line
+        else:
+            display_line = line
+
+        # If the line is Python content, append the new line, then clip the
+        # recent history to the most recent 10 clean lines, and build a single
+        # string that is the tail of the recent clean lines.
+        if included:
+            self.recent_history.append(clean_line)
+            self.recent_history = self.recent_history[-10:]
+            tail = "\n".join(self.recent_history)
+
+            # Look for the success/failure conditions in the tail
+            if self.failure_filter and self.failure_filter(tail):
+                self.success = False
+                self.log_popen.send_signal(signal.SIGINT)
+            elif self.success_filter and self.success_filter(tail):
+                self.success = True
+                self.log_popen.send_signal(signal.SIGINT)
```

Comment:
What's the purpose of sending `SIGINT` here? Why shouldn't the process finish normally?

Reply:
Primarily because on iOS and Android, it won't ever finish. iOS and Android apps don't *ever* truly exit - it's assumed that the app will start, and will persist indefinitely; on macOS, the log stream will persist.

However, we might be able to use the `stop_func` in the Android/iOS case; we're already using the stop func on the macOS case. Let me dig into this a bit.

Reply:
I taken another look at this, and I think I've found an alternate approach. Instead of sending SIGINT, we can tell the streamer to stop. The process will then fall back on the normal `subprocess.cleanup()`. We'll still potentially be terminating the process, but the logic is all contained in the subprocess module.

In the process of working this out, I was also able to build a `stop_func` implementation for iOS, so we now have a safety mechanism there, too. In normal operation, that safety mechanism won't be triggered; but if someone manages to build an app that fully crashes, the log will terminate along with the app, which brings iOS in alignment with every other platform.